### PR TITLE
fix: Fixes currentRow calculation logic in table(property pane)

### LINKED ISSUE #34346: [Bug]: `currentRow` doesn't work correctly when the table is filtered
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

I'm computing the max value of a number column in a table. To compute this I'm using the `currentRow` value. When the table is filtered, e.g., sorting by any column or making a search, the `currentRow` value is incorrect, showing the data for the row that was in that index when the table was not filtered.

Seems that `currentRow` is computed using `currentIndex`over the full data, instead of over the filtered data.

Although I found this when setting the Max value for a column, this could be happening in other validation fields.

### Steps To Reproduce

1. Create a table and fill it with data.
2. Make the columns editables.
3. Create a JS object with a method to log the row passed as parameter.
4. Set this method to one of the columns in the max validation field passing the current row as parameter, `{{ js_object.method(currentRow) }}`.
5. Filter the table sorting by a column or searching anything.
6. Click on the column previously configured with the max validation field.
7. Check in the console that the value returned for the `currentRow` is incorrect.

### Public Sample App

_No response_

### Environment

Production

### Severity

Medium (Frustrating UX)

### Issue video log

_No response_

### Version

Appsmith v1.28

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description

<ins>Problem</ins>
`currentRow` variable which is availabe in property pane of col settings - is not getting correct value during runtime if the table is sorted or filtered.

<ins>Root cause</ins> 
* We are considering `processedTableData` to get the `currentRow`.
* This property is not updated during filtering and sorting, another property `filteredTableData` is updated instead.
* We CANNOT use `filteredTableData` as it depends on `primaryColumns` property which we intend to update as well - this is leading to cyclic dependency during evaluations.

<ins>Solution</ins>
Since the problem is related to edit cases and given the constraints around using `filteredTableData` directly, we fixed the problem by adding a new property to `editableCellValue` object called `__originalIndex__`. 
* This property stores the index of the row being edited in `processedTableData`
* On top of this change, the PR adds a migration for updating the current row binding in TableWidgetV2, ensuring accurate current row calculation and improving the functionality of the widget.
* We also added unit test for migration changes.
    * Additionally, This pull request refactors the DSLs for TableWidgetV2 migration test cases to update the DSLs to separate folder, drastically reducing the file size to its core logic, improving the readability of the code. 
 * We also updated relevant test cases to account for this change.

[Testing plan](https://www.notion.so/appsmith/Issue-34346-currentRow-doesn-t-work-correctly-when-the-table-is-filtered-449225ae822c485493036599c2b19487)
[Counter part EE pr](https://github.com/appsmithorg/appsmith-ee/pull/4879)

Fixes #34346
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/10451549845>
> Commit: d1d65c6898c223bf3f6dfbfe93b8e8de214fcc7d
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=10451549845&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Mon, 19 Aug 2024 11:15:04 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [x] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Updated DSL migration process to support version 90, enhancing compatibility and robustness.
	- Introduced new migration logic for table widget data bindings, improving inline editing capabilities.
	- Enhanced validation logic for editable cells in table widgets, allowing for more dynamic data handling.
	- Added a method to discard edits in specific table cells, improving user interaction.
	- Introduced a new message constant for required fields, enhancing user feedback.

- **Bug Fixes**
	- Improved validation checks for table cells based on updated indices and values.

- **Tests**
	- Added comprehensive tests to validate migration functions related to Table Widget, ensuring all aspects function correctly post-update.
	- Enhanced test specifications for improved validation logic and coverage in table widget functionalities.

### CODE REVIEW

**Comment by rahulbarwal** [comment]
@coderabbitai review

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 1]
```diff
@@ -161,12 +167,30 @@ describe(
         "Row with id {{currentRow.id}} is not valid",
       );
 
-      table.EditTableCell(1, 4, 5, false);
-      cy.get(".bp3-popover-content").contains("Row with id 7 is not valid");
+      table.EditTableCell(1, 4, 1, false);
+      cy.get(".bp3-popover-content").contains("Row with id 2 is not valid");
+
+      cy.discardTableCellValue(4, 1);
+      agHelper.Sleep(500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:174]: please remove agHelper.Sleep(500). this creates flakiness in spec

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 2]
```diff
@@ -161,12 +167,30 @@ describe(
...
+      cy.discardTableRow(5, 1);
+
+      // Sort the id column and ensure that `currentRow` is correctly evaluated
+      table.SortColumn("id", "descending");
+      table.EditTableCell(1, 4, 1, false);
+      cy.get(".bp3-popover-content").contains("Row with id 24 is not valid");
+
+      cy.discardTableCellValue(4, 1);
+      agHelper.Sleep(500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:183]: same here remove agHelper.Sleep(500) completely

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 3]
```diff
@@ -180,14 +204,37 @@ describe(
...
+      table.EditTableCell(1, 4, 20, false);
       cy.get(".bp3-popover-content").should("not.exist");
       table.UpdateTableCell(1, 4, 8);
       cy.get(".bp3-popover-content").contains("Invalid input");
-      table.UpdateTableCell(1, 4, 7);
+      table.UpdateTableCell(1, 4, 20);
       cy.get(".bp3-popover-content").should("not.exist");
+
+      cy.discardTableCellValue(4, 1);
+      agHelper.Sleep(500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:216]: same as above

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 4]
```diff
@@ -180,14 +204,37 @@ describe(
...
+      table.SortColumn("id", "descending");
+      table.EditTableCell(1, 4, 24, false);
+      cy.get(".bp3-popover-content").should("not.exist");
+      table.UpdateTableCell(1, 4, 10);
+      cy.get(".bp3-popover-content").contains("Invalid input");
+      table.UpdateTableCell(1, 4, 24);
+      cy.get(".bp3-popover-content").should("not.exist");
+
+      cy.discardTableCellValue(4, 1);
+      agHelper.Sleep(500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:229]: same as above

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 5]
```diff
@@ -209,8 +256,8 @@ describe(
       cy.wait(1500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:256]: remove these waits as well

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 6]
```diff
@@ -237,6 +284,24 @@ describe(
...
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:284]: remove cy.wait

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 7]
```diff
@@ -237,6 +284,24 @@ describe(
...
+      cy.get(".bp3-popover-content").should("not.exist");
+      table.UpdateTableCell(1, 4, 10);
+      cy.get(".bp3-popover-content").should("not.exist");
+
+      propPane.UpdatePropertyFieldValue("Required", "{{currentRow.id == 24}}");
+      table.EditTableCell(1, 4, "");
+      cy.get(".bp3-popover-content").contains("This field is required");
+
+      cy.discardTableCellValue(4, 1);
+      agHelper.Sleep(500);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:299]: please remove sleep

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]
1. Remove all static waits please 
2. Please use locators from locators file instead of directly using them

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 1 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Done.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 2 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Done.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 3 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Done.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 5 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Done.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 4 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Removed.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 6 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: Done.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js (hunk 7 above)
Comment (rahulbarwal) [reply app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js]: removed.

**Review by riodeuno — APPROVED** [review]
Ideally, I would like to revisit the computations of `filtereTableData`, `primaryColumns` and `processedTableData` and their relationships, to see if this can be made cleaner.

However, for the purposes of Fixed mode this works. We will visit this in Anvil when necessary.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 8]
```diff
@@ -234,8 +263,24 @@ describe(
...
+      // Sort the id column and ensure that `currentRow` is correctly evaluated
+      table.SortColumn("id", "descending");
+      table.EditTableCell(1, 4, "", false);
+      cy.get(".bp3-popover-content").should("not.exist");
+      table.UpdateTableCell(1, 4, 10);
+      cy.get(".bp3-popover-content").should("not.exist");
+
+      propPane.UpdatePropertyFieldValue("Required", "{{currentRow.id == 24}}");
+      table.EditTableCell(1, 4, "");
+      cy.get(".bp3-popover-content").contains("This field is required");
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:277]: can you use this from messages file please? `This field is required` @rahulbarwal

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js [hunk 9]
```diff
@@ -234,8 +263,24 @@ describe(
...
       cy.get(".bp3-popover-content").should("not.exist");
 
-      cy.wait(1500);
-      cy.discardTableRow(5, 2);
+      table.DiscardEditRow(2, 5);
+
+      // Sort the id column and ensure that `currentRow` is correctly evaluated
+      table.SortColumn("id", "descending");
+      table.EditTableCell(1, 4, "", false);
+      cy.get(".bp3-popover-content").should("not.exist");
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/TableV2_PropertyPane_2_spec.js:271]: Please add `.bp3-popover-content` to locators @rahulbarwal

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

**Review by sagar-qa007 — APPROVED** [review]

**Review by ApekshaBhosale — APPROVED** [review]

**Pull request merged**